### PR TITLE
Adding integration test to feature controller

### PR DESCRIPTION
**Pull request by JonCrowther**
## Problem
Currently, the integration tests are a little bulky. They set up an entire Rancher instance and some even provision downstream clusters. While those are valuable from a end to end perspective, when developing they are often tricky to modify and run regularly for developers.
 
## Solution
For the controllers, the [envtest package](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) provides a library that starts a minimal control plane that allows our controllers to be registered and used for integration tests. It is lightweight, fast and requires minimal set up. While most documentation for [envtest refers to controller-runtime based](https://book.kubebuilder.io/reference/envtest.html) controllers, this PR introduces envtest for a Wrangler controller.

## Future work
- Adding these tests to CI
- Implementing envtests for a Norman based controller
- More coverage!

### CODE REVIEW

File: pkg/controllers/management/feature/suite_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+var _ = BeforeSuite(func() {
+	ctx, cancel = context.WithCancel(context.TODO())
+
+	var err error
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "..", "crds", "yaml", "generated", "management.cattle.io_features.yaml")},
+		ErrorIfCRDPathMissing: true,
+	}
+	cfg, err = testEnv.Start()
+	Expect(err).NotTo(HaveOccurred())
```
Comment (raulcabello) [inline pkg/controllers/management/feature/suite_test.go:45]: CI in drone is failing here, maybe you have to install [setup-envtest](https://github.com/rancher/fleet/blob/master/.github/scripts/run-integration-tests.sh#L9-L12) ?

File: pkg/controllers/management/feature/feature_test.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		It("adds annotation", func() {
+			f := &v3.Feature{}
+			//Eventually(func() bool {
+			//err := k8sClient.Get(ctx, types.NamespacedName{Namespace: "test-namespace", Name: "harvester-baremetal-container-workload"}, &f)
+			//var err error
+			f, _ = wranglerContext.Mgmt.Feature().Get("harvester-baremetal-container-workload", v1.GetOptions{})
+			//return err == nil
+			//}).Should(BeTrue())
+
+			Expect(f.Annotations[v3.ExperimentalFeatureKey]).To(Equal(v3.ExperimentalFeatureValue))
```
Comment (raulcabello) [inline pkg/controllers/management/feature/feature_test.go:45]: It'd be nice to add an eventually here to prevent race condition since it might take some time to update the annotation

File: pkg/controllers/management/feature/feature_handler.go [hunk 3]
```diff
@@ -96,7 +96,7 @@ func (h *handler) syncHarvesterFeature(obj *v3.Feature) error {
 
 func (h *handler) toggleHarvesterNodeDriver(harvester string) error {
 	if val := features.GetFeatureByName(harvester).Enabled(); val {
-		m, err := h.nodeDriverController.Cache().Get(harvester)
+		m, err := h.nodeDriverController.Get(harvester, metav1.GetOptions{})
```
Comment (JonCrowther) [inline pkg/controllers/management/feature/feature_handler.go:99]: This is not permanent. I'm not currently able to get the cache to work, so for now I need to change this to a full GET. DO NOT MERGE

File: pkg/apis/management.cattle.io/v3/global_types.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ type Setting struct {
 
 // +genclient
 // +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster
```
Comment (raulcabello) [inline pkg/apis/management.cattle.io/v3/global_types.go:24]: why did you need to add this? Would this break or change existing functionality?

File: pkg/crds/yaml/generated/management.cattle.io_features.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
+---
```
Comment (raulcabello) [inline pkg/crds/yaml/generated/management.cattle.io_features.yaml:1]: we don't need to generate this file since we are using `factory.BatchCreateCRDs` to register the CRDs

File: pkg/crds/yaml/generated/management.cattle.io_nodedrivers.yaml [hunk 6]
```diff
@@ -0,0 +1,120 @@
+---
```
Comment (raulcabello) [inline pkg/crds/yaml/generated/management.cattle.io_nodedrivers.yaml:1]: This file is not needed either

File: tests/controllers/feature/feature_test.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	}
+	harvesterFeature = v3.Feature{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "harvester-baremetal-container-workload",
+			Annotations: map[string]string{
+				v3.ExperimentalFeatureValue: v3.ExperimentalFeatureKey,
+			},
+		},
+	}
+	//t = true
```
Comment (raulcabello) [inline tests/controllers/feature/feature_test.go:43]: nit: remove

File: tests/controllers/feature/feature_test.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+			Name: "harvester",
+		},
+	}
+)
+
+func (s *FeatureTestSuite) SetupSuite() {
+	s.ctx, s.cancel = context.WithCancel(context.TODO())
+	// Load CRD from YAML for REST Client
+	s.testEnv = &envtest.Environment{
+		CRDDirectoryPaths: []string{
```
Comment (raulcabello) [inline tests/controllers/feature/feature_test.go:56]: This is not needed as we are using `factory.BatchCreateCRDs` to register the CRDs

File: tests/controllers/feature/feature_test.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+	}, "Feature", false)
+	assert.NoError(s.T(), err)
+	err = sc.Start(s.ctx, 1)
+	assert.NoError(s.T(), err)
+
+	s.wranglerContext.ControllerFactory.SharedCacheFactory().StartGVK(s.ctx, schema.GroupVersionKind{
+		Group:   "management.cattle.io",
+		Version: "v3",
+		Kind:    "nodedrivers",
+	})
```
Comment (raulcabello) [inline tests/controllers/feature/feature_test.go:102]: The problem is that the cache for `NodeDriver` is not initialized, and lasso is ignoring the `StartGVK` call. It retrurns [here](https://github.com/rancher/lasso/blob/master/pkg/cache/sharedinformerfactory.go#L83). We should call `ForKind` first in order to initialize the `NodeDriver` cache.

```suggestion
	_, err = s.wranglerContext.ControllerFactory.SharedCacheFactory().ForKind(schema.GroupVersionKind{
		Group:   "management.cattle.io",
		Version: "v3",
		Kind:    "NodeDriver",
	})
	assert.NoError(s.T(), err)

	err = s.wranglerContext.ControllerFactory.SharedCacheFactory().StartGVK(s.ctx, schema.GroupVersionKind{
		Group:   "management.cattle.io",
		Version: "v3",
		Kind:    "NodeDriver",
	})
```

File: tests/controllers/feature/feature_test.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	_, err = s.wranglerContext.Mgmt.NodeDriver().Update(nCopy)
+	assert.NoError(s.T(), err)
+
+	s.triggerSync("harvester")
+
+	assert.EventuallyWithT(s.T(), func(c *assert.CollectT) {
+		nd, err := s.wranglerContext.Mgmt.NodeDriver().Get("harvester", v1.GetOptions{})
+		assert.NoError(c, err)
+		assert.True(c, nd.Spec.Active)
+	}, 10*time.Second, 1*time.Second, "harvester feature has not been enabled")
```
Comment (raulcabello) [inline tests/controllers/feature/feature_test.go:134]: nit: It's be nice to move this to constants since they are used in many places

File: pkg/controllers/management/feature/feature_handler.go (hunk 3 above)
Comment (raulcabello) [reply pkg/controllers/management/feature/feature_handler.go:99]: [This](https://github.com/rancher/rancher/pull/44766#discussion_r1565393800) should fix the problem with the cache.

Comment (JonCrowther) [reply pkg/controllers/management/feature/feature_handler.go]: That did the trick!

File: tests/controllers/feature/feature_test.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	// The feature controller checks the harvester NodeDriver, so it needs to exist
+	n, err := s.wranglerContext.Mgmt.NodeDriver().Create(&v3.NodeDriver{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "harvester",
+		},
+	})
+	assert.NoError(s.T(), err)
+	assert.NotNil(s.T(), n)
+
+	// Initialize Features
```
Comment (bigkevmcd) [inline tests/controllers/feature/feature_test.go:96]: How about...

```go
// Initialize Features - this creates feature resources including the "harvester" feature.
```

It wasn't obvious to me how the feature that was being synced was being created.

File: tests/controllers/feature/feature_test.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+		assert.NoError(c, err)
+		assert.Equal(c, v3.ExperimentalFeatureValue, f.Annotations[v3.ExperimentalFeatureKey])
+	}, duration, tick, "annotation feature.cattle.io/experimental has not changed to true")
+
+	// Check that harvester feature gets set
+	assert.EventuallyWithT(s.T(), func(c *assert.CollectT) {
+		nd, err := s.wranglerContext.Mgmt.Feature().Get("harvester", v1.GetOptions{})
+		assert.NoError(c, err)
+		assert.NotNil(c, nd.Spec.Value)
+	}, duration, tick, "harvester feature has not been enabled")
```
Comment (bigkevmcd) [inline tests/controllers/feature/feature_test.go:157]: Could these two be compressed into a single assertion, not just that `ns.Spec.Value` is not `nil` but that it's true?

**Review by bigkevmcd — COMMENTED** [review]
Some good bits in here.

One big thing missing is how to run these tests.

Maybe we want a Makefile target that sets up `setup-envtest` and runs the tests?

Also, maybe a short document on how to setup your environment to run these?

(I use something like `export KUBEBUILDER_ASSETS="$(setup-envtest use 1.28.3 -p path)"` which means my editor can find it to run them (I use vim w/vim-test).

**Review by bigkevmcd — APPROVED** [review]
Thanks for the docs!

File: tests/controllers/Makefile [hunk 13]
```diff
@@ -0,0 +1,2 @@
+test:
```
Comment (raulcabello) [inline tests/controllers/Makefile:1]: nit: It'd be nice to rename this to controller-test or something like that. I would expect `make test` to run all tests not just the controller tests

**Review by raulcabello — COMMENTED** [review]
Great work! This looks much better now! Should we run these tests in CI?

**Comment by JonCrowther** [comment]
> Great work! This looks much better now! Should we run these tests in CI?

I think that should be done in a follow up PR

**Review by raulcabello — APPROVED** [review]

**Review by bigkevmcd — APPROVED** [review]
